crates/storage/provider/src/trie/mod.rs

```diff
@@ -197,6 +202,95 @@ impl<'tx, 'itx, DB: Database> DupHashDatabase<'tx, 'itx, DB> {
     }
 }
 
+/// Database wrapper implementing HashDB trait, with a read-only transaction.
+struct HashDatabase<'tx, 'itx, TX: DbTx<'itx>> {
+    tx: &'tx TX,
+    _p: PhantomData<&'itx ()>, // to suppress "unused" lifetime 'itx
+}
+
+impl<'tx, 'itx, TX> cita_trie::DB for HashDatabase<'tx, 'itx, TX>
+where
+    TX: DbTx<'itx>,
+{
+    type Error = TrieError;
+
+    fn get(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+        Ok(self.tx.get::<tables::AccountsTrie>(H256::from_slice(key))?)
+    }
+
+    fn contains(&self, key: &[u8]) -> Result<bool, Self::Error> {
+        Ok(<Self as cita_trie::DB>::get(self, key)?.is_some())
+    }
+
+    fn insert(&self, _key: Vec<u8>, _value: Vec<u8>) -> Result<(), Self::Error> {
+        // this could be avoided if cita_trie::DB was split into two traits
+        // with read and write operations respectively
+        unimplemented!("insert isn't valid for read-only transaction");
+    }
+
+    fn remove(&self, _key: &[u8]) -> Result<(), Self::Error> {
+        unimplemented!("remove isn't valid for read-only transaction");
+    }
+
+    fn flush(&self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+impl<'tx, 'itx, TX: DbTx<'itx>> HashDatabase<'tx, 'itx, TX> {
+    /// Instantiates a new Database for the accounts trie, with an existing root
+    fn from_root(tx: &'tx TX, root: H256) -> Result<Self, TrieError> {
+        tx.get::<tables::AccountsTrie>(root)?.ok_or(TrieError::MissingRoot(root))?;
+        Ok(Self { tx, _p: Default::default() })
+    }
+}
+
+/// Database wrapper implementing HashDB trait, with a read-only transaction.
+struct DupHashDatabase<'tx, 'itx, TX: DbTx<'itx>> {
+    tx: &'tx TX,
+    key: H256,
+    _p: PhantomData<&'itx ()>, // to suppress "unused" lifetime 'itx
+}
+
+impl<'tx, 'itx, TX> cita_trie::DB for DupHashDatabase<'tx, 'itx, TX>
+where
+    TX: DbTx<'itx>,
+{
+    type Error = TrieError;
+
+    fn get(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+        let mut cursor = self.tx.cursor_dup_read::<tables::StoragesTrie>()?;
+        Ok(cursor.seek_by_key_subkey(self.key, H256::from_slice(key))?.map(|entry| entry.node))
+    }
```

Comment:
cc @joshieDo @onbjerg 

the reason why I don't like the Cita Trie lib is because it doesn't let us reuse our cursors when we're walking thru the entire table, feels a bit gross to re-instantiate a cursor on every `get`